chia/wallet/util/wallet_sync_utils.py

```diff
@@ -389,3 +390,29 @@ async def fetch_header_blocks_in_range(
         assert res_h_blocks is not None
         blocks.extend([bl for bl in res_h_blocks.header_blocks if bl.height >= start])
     return blocks
+
+
+async def fetch_puzzle_solution_with_coin_info(height: uint32, coin: Coin, peer: WSChiaConnection) -> CoinSpend:
```

Comment:
I think while we're adding a new message, we should make it have a list of coins, so we can perform bulk lookups

Reply:
This might complicate the rate limiting, are you sure it's a good idea?  You can essentially make a bunch of these requests as one request and it only counts as a single request according to the rate limit.

Reply:
I'm not sure it's a good idea. but we also currently suffer from the issue of only ever having a single request outstanding (afaict) in the wallet. We can solve that by pipelining these requests, but since we would need to spawn a bunch of tasks to do that, it won't be very elegant.

I suppose making the rate limit count requests internal to messages wouldn't be very elegant either.

However, consider the case where the wallet has 100 coins it wants to look up in the same block. It would be a net win for the node to look them all up at the same time.

Maybe the trade-off is that all coins requested to be looked up has to be in the same block, i.e. a single `height` field in the message.

Reply:
> However, consider the case where the wallet has 100 coins it wants to look up in the same block. It would be a net win for the node to look them all up at the same time.

I think this is probably not very likely almost ever.  I _could_ see it happening in the case of like a bulk mint but that being a one time event means it's probably not the end of the world if the wallet slows down while processing that one block.  And I'm not sure the wallet would have the coordination to batch the request in that scenario either, it would probably end up making a bunch of requests anyways.

Reply:
Also I think we used to have a "pipeline" concept with the WalletActionStore so maybe that's something we can revisit in a more elegant way

Reply:
yeah, I'm a bit ignorant of what situations are likely to happen in real life. I don't believe any of these lookups in the toilet wallet are in the same block for instance.